Members importer handling stripe connected errors

naz opened the pull request:
These changes are meant to harden error handling for the bulk import operations and provide means to investigate possible failures. 

@allouis just wanted to hear your opinion on moving back from having `bulkAdd` on model layer for each model to dealing with bulk operations on a single module level - bulk-operations. The idea behind this move is simple - less code duplication and don't see a specific benefit of having to use models. Still need to add some error wrapping code here but think you could get the idea from changes so far ;) Let me know if you have any feedback!

naz committed:
Added error handling for failed Stripe-connected records

no issue

- Moved from using model's `bulkAdd` method to one available in bulk-operations module. The reason being we duplicate a lot of code with `bulkAdd` and that doesn't feel like it belongs to the model
- If we leave bulk operations to the importer, it would be a lot easier to extract it into separate module next
- The downside is - there's no transaction support in bulk-operations module, but with granular handling we have right now that doesn't seem necessary

naz committed:
Unified importer to use bulk-operations instead of model methods

- Duplicating same code across models didn't seem like a good idea. In addition, importer was becoming dependent on model layer which would make it harder to extract to its separate package

naz committed:
Fixed linting errors

allouis commented:
I'm thinking that it's better to have _all_ database access/modification code in the models, IMO it serves as a good separation of concerns, so that the importer wouldn't have to be changed if we were to find ways to optimise queries by their batch size or something :thinking: 

I totally get the desire to remove the duplication, as these methods are all exactly the same, but in that case, would it be worth moving these methods to the base model? 

IDK - I don't have super strong feelings about this, but my inclination is to have it in the models, especially if we're going to extract the code to another module - a model is (IMO) a good interface/contract between modules, whereas giving a module full database access seems a little too unrestricted

naz committed:
Moved bulk insert operations into models Base

- As discussed in the PR model layer seems like the most appropriate place for these operations. Follow up will include moving bulk delete operation

naz committed:
Moved bulk destroy operations into model base

naz committed:
Updated error handling and logging for bulk delete operations

- When an error happens here it's very-very unlikely scenario but still need to make sure we have proper logging in place to trace back the failure

naz committed:
Updated boolean validations for members uploads

naz commented on core/server/models/base/bulk-operations.js:
```diff
@@ -41,13 +43,20 @@ async function insert(table, data) {
         } catch (err) {
-            result.errors.push(err);
-            result.unsuccessfulIds.push(record);
+            const importError = new errors.DataImportError({
```
@allouis comparing to bulk _inserts_ we don't have special handling for whe the removing members fails. This is why such in-place handling made the most sense to me. Also, `DataImportError` was inspired by the error handling done in the regular ghost importer. 

The insert failures would be handled inside the main import loop though, with more specific error detection I think.

naz committed:
Removed date handling validation

- This is not handled on input sanitization layer with date format check in JSON schema validation

naz committed:
Added debug logs

naz committed:
Added handling for unrecognized errors

- Handling inspired by current unrecognized error handling by ghost importer -https://github.com/TryGhost/Ghost/blob/10e5d5f3d470ae201072e41213ecf708051f0a8b/core/server/data/importer/importers/data/base.js#L148-L154

naz committed:
Added details about failing record to thwon error in bulk insertions

- This allows to track and handle errors with details about which specific record caused the problem

naz committed:
Added failed key contraint handling to members_stripe_* records

- Similar handling to failed inserts for members
- Copy used for errors needs a review

naz committed:
Refactored results object to be more reabale and added debug info

naz committed:
Wrapped Stripe erros into ignition errors

- This way our logging system would know how to handle these erros
- Also added errorDetails with specific records that caused the failure

naz committed:
Added deleteMembers and insertLabels errors to logged errors

- Even though we don't return these errors to be displayed to the customer, because they are internal errors they still should be logged

naz committed:
Added additional debug logs

- Should allow easier debugging and performance profiling

naz commented:
These changes have landed in master with https://github.com/TryGhost/Ghost/commit/95044e3ba0fb76c23d7b3db2f3b2c904d7242c15
